### PR TITLE
API: send Login actions

### PR DESCRIPTION
**Pull request by AgnesToulet**
**What this PR does / why we need it**:
Send Login actions to be handled in Enterprise.

**Special notes for your reviewer:**
Some updates have been made to the `AuthenticateUser` function to have more information in the logs. I updated the authentication middleware function and the login function to keep the same behavior as before.

### CODE REVIEW

File: pkg/api/login_oauth.go
Comment (sakjur) [inline pkg/api/login_oauth.go:278]: I find it hard to follow when there are more than ~3 arguments to a function, this structure could be replaced by something like:

```suggestion
type loginError struct {
	action string
	httpStatus int
	publicMessage string
	err error
}

func (hs *HTTPServer) handleOAuthLoginError(ctx *models.ReqContext, loginError loginError) {
```

File: pkg/login/auth.go
Comment (sakjur) [inline pkg/login/auth.go:67]: Random thought: If we'd have a custom `AuthenticationError` type we would be able to send metadata about what's wrong that isn't part of the `.Error()` 🤔 That would allow for making sure that any and all authentication errors give a vague message outwards while retaining full legibility internally.

File: pkg/api/login.go
Comment (sakjur) [inline pkg/api/login.go:331]: Won't this complain for everything all the time when you're running OSS?

Comment (AgnesToulet) [reply pkg/api/login.go:331]: No, the error will be `bus.ErrHandlerNotFound` so we won't log it.

File: pkg/api/login_oauth.go
Comment (sakjur) [inline pkg/api/login_oauth.go:315]:
```suggestion
type loginInformation struct {
```

Only the inner fields have to be public 🙂

Comment (AgnesToulet) [reply pkg/api/login_oauth.go:315]: I need to access it from the `saml` package as I do pretty much the same there.

**Comment by marefr** [comment]
Would be nice to plug this in as some kind of middleware instead, but guessing that require a lot of additional work? Don't know how much of the codebase you're planning to cover, but if it's quite a lot more I would definitely consider a pluggable solution to cover this cross-cutting concern.

**Comment by sakjur** [comment]
@marefr The other part of the auditing is added as a global middleware, for login specifically we need some extra information that isn't accessible for a middleware with the current architecture.

**Comment by marefr** [comment]
@sakjur great :+1:

File: pkg/api/login.go
Comment (sakjur) [inline pkg/api/login.go:172]: Would it make sense to use defer to send this rather than defining it at all the exit points? 🤔

If we could pick up the status and error from the response object being returned instead of having to duplicate it everywhere, that'd reduce the complexity of the rest of the logic in this function quite a bit.

File: pkg/api/login_oauth.go
Comment (sakjur) [inline pkg/api/login_oauth.go:282]: Preference: When you need multiple lines for function declarations and calls in Go, it's quite common to put each variable on it's own line.

```suggestion
func syncUser(
	ctx *models.ReqContext,
	extUser *models.ExternalUserInfo,
	connect social.SocialConnector,
) (*models.User, error) {
```

File: pkg/login/auth_test.go
Comment (sakjur) [inline pkg/login/auth_test.go:30]: Is it important for this particular test to test for the value of the auth module?

**Review by sakjur — COMMENTED** [review]
:+1: I think this is starting to look quite finished, I'm a little unsure about the LoginPost function, it feels like that could benefit from being slightly refactored.

I'd also like to test running this before we merge 🙂

File: pkg/api/login.go
Comment (AgnesToulet) [reply pkg/api/login.go:172]: Good point! I searched for a better way of doing this but didn't think about defer.

File: pkg/login/auth_test.go
Comment (AgnesToulet) [reply pkg/login/auth_test.go:30]: I think it helps to understand the behavior of the function for this field.

**Review by xlson — APPROVED** [review]
Great work Agnes! The only thing that comes to find is that I think we should be asynchronous on the service side rather than in the calling code.

File: pkg/api/login.go
Comment (AgnesToulet) [reply pkg/api/login.go:168]: I agree with you that we want the audit logging to be asynchronous but it's not really what `defer` is doing and I used it for clarity mostly. (Thanks to [Emil's comment](https://github.com/grafana/grafana/pull/27249#discussion_r482098163))

Comment (xlson) [reply pkg/api/login.go:168]: Sorry, I misread the code. Then it makes perfect sense.

**Pull request merged**
